### PR TITLE
[TS] Refactor Post Display Page to Use Mock Data

### PR DESCRIPTION
**Pull request by a-shigemichi**
## Issue Number
https://github.com/okuda-seminar/Twitter-Clone/issues/533

## Implementation Summary
Changed post content display in page/[id] to use mock data.

## Particular points to check
1. Run `docker compose up`.
2. Open `localhost:3000/posts/{any}` .
3. You should see the post detail page.

スクリーンショット 2024-12-17 14 12 53

## Test
`ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx`
1. Start the frontend server (in the Twitter-Clone directory)
```
$ docker compose up -d
```
2. Enter the application container and run all test files
```
$ docker compose exec app bash
$ yarn test post-detail.spec.tsx
```
　Use `--coverage` flag to generate coverage report

## Schedule
2024/12/20

### CODE REVIEW

File: ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { render, screen, fireEvent } from "@testing-library/react";
+import { PostDetail } from "../_components/post-detail";
+
+describe("Post Detail Tests", () => {
+  test("Rendering PostDetail should success", () => {
+    render(<PostDetail />);
+  });
+  test("Follow button display in initial render", () => {
```
Comment (TakayaShirai) [inline ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx:8]: "display" -> "displays". Also, please add a blank line between each test for better readability.

```suggestion
  test("Rendering PostDetail should success", () => {
    render(<PostDetail />);
  });
  
  test("Follow button displays in initial render", () => {
```

File: ts/twitter/src/app/posts/[id]/_components/use-post-detail.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+interface usePostDetailReturn {
+  followDisplayText: string;
+  toggleFollowState: () => void;
+}
+
+export const usePostDetail = (): usePostDetailReturn => {
+  const follow = useRef<FollowState>(FollowState.UnFollow);
+  const [followDisplayText, setFollowDisplayText] = useState<string>("follow");
```
Comment (TakayaShirai) [inline ts/twitter/src/app/posts/[id]/_components/use-post-detail.ts:17]: "setFollowDisplayText" -> "setFollowButtonText"

```suggestion

export const usePostDetail = (): usePostDetailReturn => {
  const follow = useRef<FollowState>(FollowState.UnFollow);
  const [followDisplayText, setFollowButtonText] = useState<string>("follow");
```

**Review by TakayaShirai — CHANGES_REQUESTED** [review]
I could reproduce the referenced view, but I couldn't understand what "the following error message" mean you mentioned in the "Particular points to check" section. Could you clarify what it means?

Additionally, I encountered an error message in the home view as follows. I guess this is caused by changes on the backend side as I was able to reproduce the error on the main branch as well. Could you add a TODO to address this issue?

Screenshot 2024-12-17 at 12 11 18 AM (2)


Screenshot 2024-12-17 at 12 11 30 AM

File: ts/twitter/src/app/posts/[id]/_components/post-detail.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            - Hide follow button for user's own posts in post detail page. */}
+          {followDisplayText}
+        </Button>
+      </Flex>
+      <Text mt={10} fontSize="xl">
+        {post.text}
+      </Text>
+    </Box>
+  );
+};
```
Comment (ryuju0911) [inline ts/twitter/src/app/posts/[id]/_components/post-detail.tsx:30]: You implemented the logic in an unnecessarily complex way.

```suggestion
export const PostDetail = () => {
  const post: Post = mockPostCard;
  const [isFollowing, setIsFollowing] = useState<boolean>(false);
  const toggleFollowStatus = () => setIsFollowing((prev) => !prev);

  return (
    <Box key={post.id} p={8} borderWidth={1} borderRadius="md">
      <Flex>
        <Avatar size="md" name={post.user_id} mr={2} />
        <Text fontWeight="bold" fontSize="xl" mr={4} mt={2}>
          {post.user_id}
        </Text>
        <Button borderRadius="full" onClick={toggleFollowStatus}>
          {/* TODO: https://github.com/okuda-seminar/Twitter-Clone/issues/452 
            - Hide follow button for user's own posts in post detail page. */}
          {isFollowing ? "Following" : "Follow"}
        </Button>
      </Flex>
      <Text mt={10} fontSize="xl">
        {post.text}
      </Text>
    </Box>
  );
};
```

**Review by TakayaShirai — APPROVED** [review]
LGTM

File: ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { render, screen, fireEvent } from "@testing-library/react";
+import { PostDetail } from "../_components/post-detail";
+
+describe("Post Detail Tests", () => {
+  test("Rendering PostDetail should succeed", () => {
+    render(<PostDetail />);
+  });
+
+  test("Follow button displays in initial render", () => {
```
Comment (ryuju0911) [inline ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx:9]:
```suggestion
  test(`"Follow" button displays in initial render`, () => {
```

File: ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    render(<PostDetail />);
+    const followButton = screen.getByRole("button", { name: "Follow" });
+    expect(followButton).toBeInTheDocument();
+  });
+
+  test("Button text changes to following after follow button click", () => {
```
Comment (ryuju0911) [inline ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx:15]:
```suggestion
  test(`Button text changes to "Following" after clicking "Follow" button`, () => {
```

File: ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    render(<PostDetail />);
+    const followButton = screen.getByRole("button");
+
+    fireEvent.click(followButton);
+    expect(
+      screen.getByRole("button", { name: "Following" })
+    ).toBeInTheDocument();
+  });
+
+  test("Button text changes to follow after unfollow action", () => {
```
Comment (ryuju0911) [inline ts/twitter/src/app/posts/[id]/__test__/post-detail.spec.tsx:25]:
```suggestion
  test(`Button text changes to "Follow" after clicking "Following" button`, () => {
```

File: ts/twitter/src/app/posts/[id]/mocks/post-card.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { Post } from "@/lib/models/post";
+
+export const mockPostCard: Post = {
```
Comment (ryuju0911) [inline ts/twitter/src/app/posts/[id]/mocks/post-card.ts:3]: nit: It's just a post, not a post card.

**Review by ryuju0911 — CHANGES_REQUESTED** [review]

**Review by ryuju0911 — APPROVED** [review]
LGTM

**Review by ryuji0123 — APPROVED** [review]
@ryuju0911 You can merge Atsushi's PR without my review unless I explicitly require changes.

**Pull request merged**